### PR TITLE
BugFix: resolve the problem that can't get jobEnd's metrics from SQLO…

### PR DESCRIPTION
**Pull request by zhang1002**
resolve the problem that can't get jobEnd's metrics from SQLOperationListener

description:
Spark remove the listener before generating the jobEnd's data

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 1]
```diff
@@ -143,4 +143,21 @@ class ExecuteStatement(
...
+    super.cancel()
+  }
+
+  override def close(): Unit = {
+    spark.sparkContext.removeSparkListener(operationListener)
+    super.close()
+  }
+
+  override protected def onError(cancel: Boolean): PartialFunction[Throwable, Unit] = {
+
```
Comment (turboFei) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:158]: nit: remove the blank line

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala [hunk 2]
```diff
@@ -143,4 +143,21 @@ class ExecuteStatement(
       }, queryTimeout, TimeUnit.SECONDS)
     }
   }
+
+  override def cancel(): Unit = {
+    spark.sparkContext.removeSparkListener(operationListener)
```
Comment (turboFei) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:148]: is it better to define a private method for this?

```
private def removeOperationListener(): Unit = {
  if (operationListener != null) {
    spark.sparkContext.removeSparkListener(operationListener)
    operationListener = null
  }
}

```

Comment (zhang1002) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/ExecuteStatement.scala:148]: now we want to get the state that this operation has finished by SparkListenerEvent.
this PR will be revised.

**Comment by pan3793** [comment]
Ignore the JDK 11 flaky test.

**Comment by ulysses-you** [comment]
My only concern is that, the `SparkListenerSQLExecutionEnd` might be dropped by Spark then we can not receive this event which can cause the memory leak.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 3]
```diff
@@ -59,6 +63,10 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
     if (sameGroupId(jobStart.properties)) {
       val jobId = jobStart.jobId
       val stageSize = jobStart.stageInfos.size
+      val execId = jobStart.properties.getProperty("spark.sql.execution.id")
+      if (executionId.isEmpty) {
+        executionId = Some(execId.toLong)
+      }
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:69]: if (executionId.isEmpty) {
        executionId = Option(jobStart.properties.getProperty("spark.sql.execution.id")).map(_.toLong)
}

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 4]
```diff
@@ -107,4 +115,26 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
   override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = activeStages.synchronized {
     if (activeStages.contains(taskEnd.stageId)) super.onTaskEnd(taskEnd)
   }
+
+  // here we will get two kind of event
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:119]: Here we will get two kind of events

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 5]
```diff
@@ -107,4 +115,26 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
...
+  //      - SparkListenerSQLExecutionStart: from this event, you can get physicPlan and logicalPlan
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:120]: - SparkListenerSQLExecutionStart

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 6]
```diff
@@ -107,4 +115,26 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
...
+  //      - SparkListenerSQLExecutionEnd
+  // 1. get the event: sparkListenerSQLExecutionStart, and get its execution_id.
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:122]: sparkListenerSQLExecutionStart -> SparkListenerSQLExecutionStart

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 7]
```diff
@@ -107,4 +115,26 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
...
+  //    then store it in executionToOperationMap, key is execution_id and value is "empty"
+  // 2. when we get the event: SparkListenerJobStart,
+  //    we need to get its execution_id from its properties and its operation_id.
+  //    then store them in executionToOperationMap,
+  //    key is execution_id and value is operation_id
+  // 3. get the event: SparkListenerSQLExecutionEnd, and get its execution_id.
+  //    then remove the data from executionToOperationMap which key is execution_id
+  //    and remove the listener from sparkContext
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:130]: the comment here is not related to the method

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala [hunk 8]
```diff
@@ -107,4 +115,26 @@ class SQLOperationListener(operation: Operation) extends StatsReportListener wit
...
+  override def onOtherEvent(event: SparkListenerEvent): Unit = {
+    event match {
+      case sqlExecutionEnd: SparkListenerSQLExecutionEnd =>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SQLOperationListener.scala:133]:
```
case sqlExecutionEnd: SparkListenerSQLExecutionEnd if (executionId.contains(sqlExecutionEnd.executionId) =>
```

**Comment by yaooqinn** [comment]
> My only concern is that, the `SparkListenerSQLExecutionEnd` might be dropped by Spark then we can not receive this event which can cause the memory leak.

we can remove it in `cleanup` then both `cancel` and `close` can ensure to clean the listener

**Review by ulysses-you — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master for v1.3.0

**Pull request closed without merging**